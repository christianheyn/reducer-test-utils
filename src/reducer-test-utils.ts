/**
 *
 * @example
 * ```
 *  const initialState = 1;
 *
 *  pipeActions(reducer,
 *    { type: ADD_1 },
 *    (state) => expect(state).toEqual(2),
 *    { type: ADD_N, payload: 3 },
 *    { type: ADD_1 },
 *    { type: SUB_1 },
 *    (state) => expect(state).toEqual(5),
 *  )(initialState);
 * ```
 */
export const pipeActions = <STATE, ACTION>(
    reducer: (state: STATE, action: ACTION) => STATE,
    ...actions: (ACTION | ((state: STATE) => void))[]
) => (state: STATE): STATE => {
    const result: STATE = actions.reduce((acc, x) => {
        if (x instanceof Function) {
            x(acc)
            return acc
        }
        return reducer(acc, x)
    }, state)

    return result
}

export const stateChanged = <STATE>(
    pipeActionsResult: (state: STATE) => STATE,
    initialState: STATE
): boolean => {
    const stateAfterActions = pipeActionsResult(initialState)
    return (
        JSON.stringify(stateAfterActions, undefined, 0) !==
        JSON.stringify(initialState, undefined, 0)
    )
}
