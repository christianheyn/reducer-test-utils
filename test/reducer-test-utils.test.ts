import { pipeActions } from '../src/reducer-test-utils'

describe('pipeActions', () => {
    it('works with multiple actions', () => {
        enum Actions {
            ADD_1 = 'ADD_1',
            ADD_N = 'ADD_N',
            SUB_2 = 'SUB_2'
        }

        function reducer(state: number, action: { type: Actions; payload?: number }) {
            switch (action.type) {
                case Actions.ADD_1:
                    return state + 1
                case Actions.ADD_N:
                    if (typeof action.payload === 'number') {
                        return state + action.payload
                    }
                    return state
                case Actions.SUB_2:
                    return state - 2
                default:
                    return state
            }
        }

        const actionPipe = pipeActions(
            reducer,
            { type: Actions.ADD_1 },
            state => expect(state).toEqual(2),
            { type: Actions.ADD_N, payload: 3 },
            { type: Actions.ADD_1 },
            state => expect(state).toEqual(6),
            state => expect(state).toMatchSnapshot(),
            { type: Actions.SUB_2 },
            function(state) {
                expect(state).toEqual(4)
            }
        )

        expect(actionPipe(1)).toEqual(4)
    })

    it('works with single action', () => {
        enum Actions {
            ADD_1 = 'ADD_1',
            ADD_N = 'ADD_N',
            SUB_2 = 'SUB_2'
        }

        function reducer(state: number, action: { type: Actions; payload?: number }) {
            switch (action.type) {
                case Actions.ADD_1:
                    return state + 1
                case Actions.ADD_N:
                    if (typeof action.payload === 'number') {
                        return state + action.payload
                    }
                    return state
                case Actions.SUB_2:
                    return state - 2
                default:
                    return state
            }
        }

        const actionPipe = pipeActions(reducer, { type: Actions.ADD_1 })

        expect(actionPipe(1)).toEqual(2)
    })

    it('works with every propName for payload', () => {
        enum Actions {
            ADD_1 = 'ADD_1',
            ADD_N = 'ADD_N',
            SUB_2 = 'SUB_2'
        }

        function reducer(state: number, action: { type: Actions; n?: number }) {
            switch (action.type) {
                case Actions.ADD_1:
                    return state + 1
                case Actions.ADD_N:
                    if (typeof action.n === 'number') {
                        return state + action.n
                    }
                    return state
                case Actions.SUB_2:
                    return state - 2
                default:
                    return state
            }
        }

        const actionPipe = pipeActions(
            reducer,
            { type: Actions.ADD_1 },
            state => expect(state).toEqual(2),
            { type: Actions.ADD_N, n: 3 },
            { type: Actions.ADD_1 },
            state => expect(state).toEqual(6),
            { type: Actions.SUB_2 },
            function(state) {
                expect(state).toEqual(4)
            }
        )

        expect(actionPipe(1)).toEqual(4)
    })
})
