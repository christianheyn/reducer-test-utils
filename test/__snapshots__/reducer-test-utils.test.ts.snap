// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pipeActions works with multiple actions 1`] = `6`;
